delay instantiation of `const_results`

aviatesk opened the pull request:
Might give a minor memory optimization.

@nanosoldier `runbenchmarks("inference", vs=":master")`

aviatesk committed:
delay instantiation of `const_results`

Might give a minor memory optimization.

nanosoldier commented:
[Your job](https://github.com/JuliaLang/julia/pull/51925) failed.

aviatesk commented:
@nanosoldier `runbenchmarks("inference", vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/51925#issuecomment-1784018122) has completed - no performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/250a0de_vs_d5c30d8/report.md).

aviatesk commented:
@nanosoldier `runbenchmarks("inference", vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/51925#issuecomment-1784682134) has completed - no performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/c6195c9_vs_ca5b4c5/report.md).

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/51925#issuecomment-1784733537) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/c052e0f_vs_55b0729/report.md).

aviatesk committed:
inference: use `const_results` even if there are unanalyzed matches

Now the length of `const_results` matches the number of matched methods
always if initialized, so there is no need to restrict the use of it
even in the presence of unanalyzed calls.

aviatesk commented:
@nanosoldier `runbenchmarks("inference", vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/51925#issuecomment-1784872706) has completed - no performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/d5d78c1_vs_55b0729/report.md).
